File: python/packages/core/src/matimo/mcp/tool_converter.py
"""
MCP tool converter — converts Matimo parameter schemas to MCP JSON Schema.
Mirrors: packages/core/src/mcp/tool-converter.ts
"""
from __future__ import annotations

import re
from typing import Any

from matimo.core.models import Parameter, ToolDefinition

# Auth-related parameter name patterns.
# Parameters matching these are excluded from the MCP input schema
# because they are injected server-side by the secret resolver.
_AUTH_PATTERNS = frozenset(
    ["token", "key", "secret", "password", "credential", "auth", "bearer"]
)


def _is_auth_parameter(name: str) -> bool:
    """
    Check if a parameter name looks like a secret/auth parameter.

    Normalises camelCase to segments first (e.g. apiKey → ['api', 'key']),
    then splits on word separators (_ - .) and checks each segment for an
    exact match against _AUTH_PATTERNS, preventing false positives such as
    "monkey" matching "key" or "author" matching "auth".
    """
    # Convert camelCase → snake_case (apiKey → api_key)
    snake = re.sub(r"([a-z])([A-Z])", r"\1_\2", name).lower()
    segments = [s for s in re.split(r"[_\-.]+", snake) if s]
    return any(segment in _AUTH_PATTERNS for segment in segments)


def convert_parameters_to_mcp_schema(
    parameters: dict[str, Parameter],
) -> dict[str, Any]:
    """
    Convert a Matimo parameters dict to a JSON Schema object suitable for
    MCP inputSchema. Auth parameters are excluded — they are injected server-side.

    Returns a JSON Schema of type: object with properties + required list.
    """
    properties: dict[str, Any] = {}
    required: list[str] = []

    for name, param in parameters.items():
        # Skip auth parameters — they are injected by the MCP server
        if _is_auth_parameter(name):
            continue
        properties[name] = _parameter_to_json_schema(param)
        if param.required:
            required.append(name)

    schema: dict[str, Any] = {"type": "object", "properties": properties}
    if required:
        schema["required"] = required
    return schema


def derive_tool_annotations(tool: ToolDefinition) -> dict[str, bool]:
    """
    Derive MCP standard tool annotations directly from execution type and
    HTTP method. Mirrors deriveToolAnnotations() in tool-converter.ts.

    More precise than projecting through the aggregate risk *level* string,
    since e.g. GET and DELETE can share a risk tier on some tools while
    having opposite readOnlyHint/destructiveHint values.

    `requires_approval=True` always forces destructiveHint to True, since
    that flag exists precisely to flag operations a human should confirm.
    """
    exec_type = tool.execution.type

    if exec_type in ("function", "command"):
        annotations = {
            "readOnlyHint": False,
            "destructiveHint": True,
            "idempotentHint": False,
            "openWorldHint": True,
        }
    elif exec_type == "http":
        method = tool.execution.method.upper()  # type: ignore[union-attr]
        if method == "GET":
            annotations = {
                "readOnlyHint": True,
                "destructiveHint": False,
                "idempotentHint": True,
                "openWorldHint": True,
            }
        elif method == "PUT":
            annotations = {
                "readOnlyHint": False,
                "destructiveHint": False,
                "idempotentHint": True,
                "openWorldHint": True,
            }
        elif method == "DELETE":
            annotations = {
                "readOnlyHint": False,
                "destructiveHint": True,
                "idempotentHint": True,
                "openWorldHint": True,
            }
        else:
            # POST, PATCH, and any other write method: not idempotent by default
            annotations = {
                "readOnlyHint": False,
                "destructiveHint": False,
                "idempotentHint": False,
                "openWorldHint": True,
            }
    else:
        # Unknown execution type — treat conservatively (matches
        # _classify_automatic_risk's HIGH fallback for unrecognized types).
        annotations = {
            "readOnlyHint": False,
            "destructiveHint": True,
            "idempotentHint": False,
            "openWorldHint": True,
        }

    if tool.requires_approval:
        annotations = {**annotations, "destructiveHint": True}

    return annotations


def humanize_tool_name(name: str) -> str:
    """
    Convert a snake_case tool name into a human-readable title for MCP
    clients, e.g. `slack_get_channel_history` -> `Slack Get Channel History`.
    Mirrors humanizeToolName() in tool-converter.ts — only the first
    character of each word is uppercased, the remainder is left as-is.
    """
    return " ".join(word[0].upper() + word[1:] for word in name.split("_") if word)


def tool_to_mcp_registration(tool: ToolDefinition) -> dict[str, Any]:
    """
    Build the full MCP tool registration metadata from a ToolDefinition.
    Mirrors toolToMcpRegistration() in tool-converter.ts.

    Returns a dict with title, description, inputSchema, and annotations.
    Tools with requires_approval get an extra ``_matimo_approved`` parameter
    so clients can confirm destructive operations.
    """
    schema = convert_parameters_to_mcp_schema(tool.parameters or {})

    # Tools with requires_approval need the _matimo_approved parameter in
    # the MCP schema so clients can confirm destructive operations.
    if tool.requires_approval:
        schema.setdefault("properties", {})["_matimo_approved"] = {
            "type": "boolean",
            "description": (
                "Set to true to confirm execution of this approval-required tool"
            ),
        }

    return {
        "title": humanize_tool_name(tool.name),
        "description": tool.description or tool.name,
        "inputSchema": schema,
        "annotations": derive_tool_annotations(tool),
    }


def _parameter_to_json_schema(param: Parameter) -> dict[str, Any]:
    """Convert a single Matimo Parameter to a JSON Schema fragment."""
    schema: dict[str, Any] = {"type": param.type.value}

    if param.description:
        schema["description"] = param.description
    if param.enum:
        schema["enum"] = param.enum
    if param.default is not None:
        schema["default"] = param.default

    if param.type.value == "array":
        if param.items:
            schema["items"] = _parameter_to_json_schema(param.items)
        else:
            # Some MCP clients/extensions require `items` to be present
            # for array-typed parameters. When the tool doesn't specify an
            # element type, provide an empty schema so validators accept it.
            schema["items"] = {}

    if param.type.value == "object" and param.properties:
        nested_props: dict[str, Any] = {}
        nested_required: list[str] = []
        for k, v in param.properties.items():
            nested_props[k] = _parameter_to_json_schema(v)
            if v.required:
                nested_required.append(k)
        schema["properties"] = nested_props
        if nested_required:
            schema["required"] = nested_required

    return schema

File: typescript/packages/core/src/mcp/tool-converter.ts
/**
 * Tool Schema Converter
 *
 * Converts Matimo ToolDefinition parameters to MCP-compatible Zod schemas.
 * The MCP SDK's registerTool() accepts { [key]: ZodType } as inputSchema.
 * Reuses the same parameterToZod logic as the LangChain integration.
 */

import { z } from 'zod';
import type { Parameter } from '../core/types.js';
import type { ToolDefinition } from '../core/schema.js';

/**
 * Convert a single Matimo Parameter to a Zod schema.
 * Handles: string, number, boolean, array, object, enum, defaults, optionals.
 */
export function parameterToZod(param: Parameter): z.ZodType<unknown> {
  let schema: z.ZodType<unknown>;

  // Handle enum constraints
  if (param.enum && param.enum.length > 0) {
    if (param.enum.length === 1) {
      // z.union requires at least 2 schemas — handle single-value enum explicitly
      schema = z.literal(param.enum[0]);
    } else {
      const enumSchemas = param.enum.map((value) => z.literal(value));
      schema = z.union(enumSchemas as unknown as [z.ZodTypeAny, ...z.ZodTypeAny[]]);
    }
  } else {
    switch (param.type) {
      case 'string':
        schema = z.string();
        break;
      case 'number':
        schema = z.number();
        break;
      case 'boolean':
        schema = z.boolean();
        break;
      case 'array': {
        const itemSchema = param.items ? parameterToZod(param.items) : z.unknown();
        schema = z.array(itemSchema);
        break;
      }
      case 'object': {
        if (param.properties) {
          const props: Record<string, z.ZodType<unknown>> = {};
          for (const [key, prop] of Object.entries(param.properties)) {
            props[key] = parameterToZod(prop);
          }
          schema = z.object(props);
        } else {
          schema = z.record(z.string(), z.unknown());
        }
        break;
      }
      default:
        schema = z.unknown();
    }
  }

  // Add description
  if (param.description) {
    schema = schema.describe(param.description);
  }

  // Make optional before applying default.
  // Order matters: .optional().default(val) produces ZodDefault(ZodOptional(...)),
  // so parse(undefined) triggers the default. Reversing the order wraps ZodDefault
  // in ZodOptional, causing undefined to be absorbed before the default is reached.
  if (!param.required) {
    schema = schema.optional();
  }

  // Apply default after optional so parse(undefined) returns the default value.
  if (param.default !== undefined) {
    schema = schema.default(param.default);
  }

  return schema;
}

/**
 * Auth-related parameter name patterns.
 * Parameters matching these are excluded from the MCP input schema
 * because they are injected server-side by the secret resolver.
 */
const AUTH_PATTERNS = ['token', 'key', 'secret', 'password', 'credential', 'auth', 'bearer'];

/**
 * Check if a parameter name looks like a secret/auth parameter.
 * Normalises camelCase to segments first (e.g. apiKey → ['api','key']),
 * then splits on word separators (_ - .) and checks each segment for an
 * exact match against AUTH_PATTERNS, preventing false positives such as
 * "monkey" matching "key" or "author" matching "auth".
 */
function isAuthParameter(paramName: string): boolean {
  const segments = paramName
    .replace(/([a-z])([A-Z])/g, '$1_$2') // camelCase → snake_case (apiKey → api_Key)
    .toLowerCase()
    .split(/[_\-.]+/)
    .filter(Boolean);
  return segments.some((segment) => AUTH_PATTERNS.includes(segment));
}

/**
 * Convert a ToolDefinition's parameters to MCP inputSchema format.
 * Excludes auth parameters — those are injected server-side.
 *
 * @returns A plain object mapping param names to Zod types,
 *          which is what MCP SDK's registerTool() expects for inputSchema.
 */
export function convertParametersToMcpSchema(
  parameters: Record<string, Parameter>
): Record<string, z.ZodTypeAny> {
  const schema: Record<string, z.ZodTypeAny> = {};

  for (const [name, param] of Object.entries(parameters)) {
    // Skip auth parameters — they are injected by the MCP server
    if (isAuthParameter(name)) {
      continue;
    }
    schema[name] = parameterToZod(param) as z.ZodTypeAny;
  }

  return schema;
}

/**
 * MCP standard tool-annotation hints (readOnlyHint, destructiveHint,
 * idempotentHint, openWorldHint — see the MCP spec's ToolAnnotations).
 * Defined locally, matching the shape of @modelcontextprotocol/sdk's
 * ToolAnnotations, rather than imported — the rest of mcp/ only ever
 * loads the SDK dynamically (see mcp-server.ts), so this module avoids
 * taking a static compile-time dependency on its type exports.
 */
export interface McpToolAnnotations {
  readOnlyHint: boolean;
  destructiveHint: boolean;
  idempotentHint: boolean;
  openWorldHint: boolean;
}

/**
 * Derive MCP standard tool annotations directly from execution type and
 * HTTP method — more precise than projecting through the aggregate risk
 * *level* string, since e.g. GET and DELETE can share a risk tier on some
 * tools while having opposite readOnlyHint/destructiveHint values.
 *
 * `requires_approval: true` always forces destructiveHint to true, since
 * that flag exists precisely to flag operations a human should confirm.
 */
export function deriveToolAnnotations(tool: ToolDefinition): McpToolAnnotations {
  const exec = tool.execution;
  let annotations: McpToolAnnotations;

  if (exec.type === 'function' || exec.type === 'command') {
    // Arbitrary code / shell execution: not read-only, treat as destructive
    // and non-idempotent (matches classifyAutomaticRisk's critical/high tiers).
    annotations = {
      readOnlyHint: false,
      destructiveHint: true,
      idempotentHint: false,
      openWorldHint: true,
    };
  } else if (exec.type === 'http') {
    switch (exec.method.toUpperCase()) {
      case 'GET':
        annotations = {
          readOnlyHint: true,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: true,
        };
        break;
      case 'PUT':
        annotations = {
          readOnlyHint: false,
          destructiveHint: false,
          idempotentHint: true,
          openWorldHint: true,
        };
        break;
      case 'DELETE':
        annotations = {
          readOnlyHint: false,
          destructiveHint: true,
          idempotentHint: true,
          openWorldHint: true,
        };
        break;
      default:
        // POST, PATCH, and any other write method: not idempotent by default
        annotations = {
          readOnlyHint: false,
          destructiveHint: false,
          idempotentHint: false,
          openWorldHint: true,
        };
    }
  } else {
    // Unknown execution type — treat conservatively (matches
    // classifyAutomaticRisk's 'high' fallback for unrecognized types).
    annotations = {
      readOnlyHint: false,
      destructiveHint: true,
      idempotentHint: false,
      openWorldHint: true,
    };
  }

  if (tool.requires_approval === true) {
    annotations = { ...annotations, destructiveHint: true };
  }

  return annotations;
}

/**
 * Convert a snake_case tool name into a human-readable title for MCP
 * clients, e.g. `slack_get_channel_history` → `Slack Get Channel History`.
 */
export function humanizeToolName(name: string): string {
  return name
    .split('_')
    .filter(Boolean)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}

/**
 * Build the full MCP tool registration metadata from a ToolDefinition.
 *
 * @returns Object ready for server.registerTool(name, metadata, handler)
 */
export function toolToMcpRegistration(tool: ToolDefinition): {
  title: string;
  description: string;
  inputSchema: Record<string, z.ZodTypeAny>;
  annotations: McpToolAnnotations;
} {
  const schema = convertParametersToMcpSchema(tool.parameters || {});

  // Tools with requires_approval need the _matimo_approved parameter in
  // the MCP schema so clients can confirm destructive operations.
  if (tool.requires_approval) {
    schema._matimo_approved = z
      .boolean()
      .optional()
      .describe('Set to true to confirm execution of this approval-required tool');
  }

  return {
    title: humanizeToolName(tool.name),
    description: tool.description || tool.name,
    inputSchema: schema,
    annotations: deriveToolAnnotations(tool),
  };
}

/**
 * Extract auth placeholder names from a tool's execution config.
 * These are the env var names the tool needs (e.g., SLACK_BOT_TOKEN, GITHUB_TOKEN).
 * Used by `matimo mcp setup` to generate config templates.
 */
export function extractAuthPlaceholders(tool: ToolDefinition): string[] {
  const placeholders: string[] = [];
  // Use \w+ (word chars only) instead of [^}]+ to avoid polynomial backtracking (ReDoS) on
  // uncontrolled input — placeholder names are always alphanumeric identifiers anyway.
  const placeholderRegex = /\{(\w+)\}/g;
  const execution = tool.execution;

  const scanString = (str: string) => {
    placeholderRegex.lastIndex = 0; // reset before each scan to avoid stale lastIndex from /g flag
    let match;
    while ((match = placeholderRegex.exec(str)) !== null) {
      const name = match[1];
      if (isAuthParameter(name) && !placeholders.includes(name)) {
        placeholders.push(name);
      }
    }
  };

  const scanObject = (obj: unknown) => {
    if (typeof obj === 'string') {
      scanString(obj);
    } else if (obj && typeof obj === 'object') {
      for (const value of Object.values(obj as Record<string, unknown>)) {
        scanObject(value);
      }
    }
  };

  // Scan URL, headers, body, query_params
  if ('url' in execution) scanString(execution.url);
  if ('headers' in execution && execution.headers) scanObject(execution.headers);
  if ('body' in execution && execution.body) scanObject(execution.body);
  if ('query_params' in execution && execution.query_params) scanObject(execution.query_params);
  if ('args' in execution && execution.args) {
    for (const arg of execution.args) {
      scanString(arg);
    }
  }
  if ('command' in execution) scanString(execution.command);

  return placeholders;
}
